RavenDB-19769 improve shrading import performance

karmeli87 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19769

### Additional description

low hanging fruits, for review.

### Type of change

- Optimization

### How risky is the change?

- Moderate 
- 
### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- It has been verified by manual testing

### Testing by RavenDB QA team

- This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). 

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

karmeli87 committed:
RavenDB-19769 avoid reading the sharding state for every document put

karmeli87 committed:
RavenDB-19769 make FlushAsync, async all the way

karmeli87 committed:
RavenDB-19769 orchestrator should not compress

karmeli87 committed:
RavenDB-19769 the gzip stream for the shard should not compress

karmeli87 committed:
RavenDB-19769 generate new context every 16MB so we will not accumulate too much cached items

ppekrol commented on src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs:
```diff
@@ -162,8 +164,21 @@ protected ShardedActions(ShardedDatabaseContext databaseContext, Dictionary<int,
+            public JsonOperationContext GetContextForNewCompareExchangeValue() => GetContextForNewDocument();
+            public JsonOperationContext GetContextForNewDocument()
+            {
+                if (_context.AllocatedMemory > 16 * 1024 * 1024)
```
This looks suspicious TBH, why we would want to do that?

karmeli87 replied:
The problem here was that if we use the same context we will end up in this situation:
![high-mem](https://issues.hibernatingrhinos.com/api/files/74-283608?sign=MTY3NDE3MjgwMDAwMHwyMC04MTl8NzQtMjgzNjA4fGxyRmJXTFJJTG03NWpGZVo1WVAtZi02RjVs%0D%0ARm9ybW1EWVZGamNQM1JtYncNCg%0D%0A&updated=1673674640121)

Because we will not dispose the modifier here:
https://github.com/ravendb/ravendb/blob/ca9ae26f52f633118c0bede47e5aeb9b8b5da640/src/Raven.Server/Smuggler/Documents/StreamSource.cs#L1337

This is probably not the best solution, and we would to want to renew/reset the `BlittableMetadataModifier`, but I will let aviv to figure this out :)

ayende replied:
Why aren't we disposing it?

aviv86 replied:
we are not disposing it here because after renewing the context, the old context is used [here](https://github.com/ravendb/ravendb/blob/v5.4/src/Raven.Server/Smuggler/Documents/StreamSource.cs#L1452) in `modifier.Dispose()` which releases the allocations from old context

ppekrol commented on src/Sparrow/Json/AsyncBlittableJsonTextWriter.cs:
```diff
@@ -61,24 +61,24 @@ public async ValueTask WriteStreamAsync(Stream stream, CancellationToken token =
 
-            FlushInternal(); // this is OK, because inner stream is a MemoryStream
-            return FlushAsync(token);
+            await FlushInternalAsync().ConfigureAwait(false); // this is OK, because inner stream is a MemoryStream
```
does this comment is still valid?
FlushInternalAsync returns ValueTask?

karmeli87 replied:
This turned out to be irrelevant to the performance. 
We can keep the `async` here or revert this change.

ppekrol replied:
Revert, lets avoid additional async machinery

aviv86 replied:
reverted

aviv86 committed:
RavenDB-19769 : revert changes in AsyncBlittableJsonTextWriter

ayende commented on src/Raven.Server/Smuggler/Documents/MultiShardedDestination.cs:
```diff
@@ -81,7 +82,7 @@ private async Task PrepareShardStreamDestination(Dictionary<int, StreamDestinati
             holders[shardNumber].InputStream = stream;
             holders[shardNumber].ContextReturn = _handler.ContextPool.AllocateOperationContext(out JsonOperationContext context);
-            var destination = new StreamDestination(stream, context, _source);
+            var destination = new StreamDestination(stream, context, _source, CompressionLevel.NoCompression);
```
This has to be configurable.

aviv86 replied:
done

aviv86 committed:
RavenDB-19769 : dispose previous context

aviv86 committed:
RavenDB-19769 : make the compression level of import streams that orchestrator sends to shards configurable

aviv86 committed:
RavenDB-19769 : fix restore and prefixed sharding failing tests
- set ShardingConfiguration in ShardedDocumentDatabase from smuggler record when doing sharded restore
- keep the Prefixed configuration when generating sharded database record from parent record
- test valid bucket stats for prefixed buckets after restore

karmeli87 commented on src/Raven.Server/Config/Categories/ShardingConfiguration.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        [Description("The compression level to use when sending import strems to shards during smuggler import")]
+        [DefaultValue(CompressionLevel.NoCompression)]
+        [ConfigurationEntry("Import.CompressionLevel", ConfigurationEntryScope.ServerWideOrPerDatabase)]
```
`Sharding.Import.CompressionLevel`

aviv86 replied:
fixed

karmeli87 commented on src/Raven.Server/Config/Categories/ShardingConfiguration.cs:
```diff
@@ -0,0 +1,15 @@
+    [ConfigurationCategory(ConfigurationCategoryType.Sharding)]
+    public class ShardingConfiguration : ConfigurationCategory
+    {
+        [Description("The compression level to use when sending import strems to shards during smuggler import")]
```
streams

aviv86 replied:
thanks, fixed

karmeli87 commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -71,6 +71,7 @@
 using Voron.Data.Tables;
 using Voron.Impl;
 using Constants = Raven.Client.Constants;
+using ShardingConfiguration = Raven.Client.ServerWide.Sharding.ShardingConfiguration;
```
revert

aviv86 replied:
done

karmeli87 commented on src/Raven.Server/ServerWide/RawDatabaseRecord.cs:
```diff
@@ -238,7 +238,6 @@ public RawDatabaseRecord GetShardedDatabaseRecord(int shardNumber)
             sharding.Modifications = new DynamicJsonValue(sharding)
             {
                 [nameof(ShardingConfiguration.BucketRanges)] = null,
```
I think we need to remove the `BucketRanges` here, otherwise this will be striped and we wouldn't able to tell the correct shard?

aviv86 replied:
I don't think we need the `BucketRanges`, we only use `ShardedDocumentDatabase.ShardingConfiguration` for updating `BucketAndEtag` indexes and `BucketStats` tree.
so we don't care about shard number, just the bucket number - and `Prefixed` is enough for that

aviv86 committed:
RavenDB-19769 : address PR comments
